### PR TITLE
Enabling WebHooks from console will add Generic webhook

### PR DESCRIPTION
**Pull request by jhadvig**
@jwforres PTAL

fixes https://github.com/openshift/origin/issues/8536

### CODE REVIEW

File: assets/app/scripts/controllers/edit/buildConfig.js [hunk 1]
```diff
@@ -633,21 +633,30 @@ angular.module('openshiftConsole')
...
         if (webhooks.length === 0) {
-          webhooks.push({
-            github: {
-              secret: ApplicationGenerator._generateSecret()
-            },
-            type: "GitHub"
-          });
+          webhooks.push(createWebhook("GitHub"));
+          webhooks.push(createWebhook("Generic"));
+        } else if (webhooks.length === 1) {
```
Comment (jwforres) [inline assets/app/scripts/controllers/edit/buildConfig.js]: hmm, this makes me think the editor should really have these as separate options.  if i choose to go in and delete one of them in the yaml editor, then i dont want it to add it back any time I use the GUI editor.

Comment (jhadvig) [inline assets/app/scripts/controllers/edit/buildConfig.js]: Well thats true, but I remember that while working on BC editor we decided to go just with one checkbox. The point was (if I remember correctly) that it doesn't make sense to create just Generic webhook.
But I see your point, so if anything I can update the logic to distinguish between those two webhooks, since now we dont.

**Comment by jhadvig** [comment]
@jwforres I've updated the PR with two separate checkbox, each for given webhook type. The only thing I'm not fully sure of are the tooltip titles (I've only update the one for the GitHub webhook), if its ok to leave them as they are.
PTAL

File: assets/app/views/edit/build-config.html [hunk 2]
```diff
@@ -473,17 +473,32 @@
...
+                                </label>
+                              </div>
+
+                              <div class="checkbox">
+                                <label>
+                                  <input type="checkbox" ng-model="triggers.present.genericWebhook"/>
+                                  Enable the Generic webhook build trigger
+                                  <span class="help action-inline">
+                                    <a href>
+                                      <i class="pficon pficon-help" data-toggle="tooltip" aria-hidden="true" data-original-title="The source repository must be configured to use the webhook to trigger a build when source is committed.">
```
Comment (jwforres) [inline assets/app/views/edit/build-config.html]: maybe just use the text from the docs: "Generic webhooks can be invoked from any system capable of making a web request."  although I might rephrase it to something like "A generic webhook can be triggered by any system capable of making a web request."

File: assets/app/scripts/controllers/edit/buildConfig.js [hunk 3]
```diff
@@ -633,21 +635,31 @@ angular.module('openshiftConsole')
     };
 
     $scope.updateTriggers = function() {
+
+      // Helper for checking webhook objects in the buildConfig spec and creating if missing.
+      function reuseOrCreateWebhook(type) {
+        var webhook = _.find($scope.buildConfig.spec.triggers, function(obj) { return obj.type === type})
```
Comment (jwforres) [inline assets/app/scripts/controllers/edit/buildConfig.js]: can you have more than one Github webhook or more than one Generic webhook?  we should make sure the editor isn't stripping out any additional ones

cc @bparees

**Comment by jhadvig** [comment]
+1 for the second

File: assets/app/views/edit/build-config.html (hunk 2 above)
Comment (jhadvig) [inline assets/app/views/edit/build-config.html]: +1 for the second

**Comment by bparees** [comment]
> can you have more than one Github webhook or more than one Generic webhook? we should make sure the editor isn't stripping out any additional ones

@jwforres yeah @jhadvig and i discussed...  today nothing prevents you from defining multiple, but our code will effectively ignore the additional ones, so we're going to add validation that prevents it.  For now i think it's reasonable for this PR to assume there's only one per type.

**Comment by jhadvig** [comment]
@bparees @jwforres https://github.com/openshift/origin/issues/8551

**Comment by liggitt** [comment]
> today nothing prevents you from defining multiple, but our code will effectively ignore the additional ones, so we're going to add validation that prevents it. For now i think it's reasonable for this PR to assume there's only one per type.

I don't think we should remove the ability to have multiple webhooks of the same type... there is value in being able to give different repos/integrators their own secret (especially for the generic webhook)

**Comment by bparees** [comment]
@liggitt to be clear the capability doesn't actually exist today, just the misleading appearance of such a capability.  but I am convinced by your scenario that the right fix is to properly handle the multiple triggers, rather than explicitly enforce the limit to 1.

**Comment by jhadvig** [comment]
> I am convinced by your scenario that the right fix is to properly handle the multiple triggers, rather than explicitly enforce the limit to 1.

+1

**Comment by jhadvig** [comment]
@jwforres I've update the PR so it wont strip out any additional triggers of one type. PTAL

File: assets/app/scripts/controllers/edit/buildConfig.js [hunk 4]
```diff
@@ -633,21 +635,31 @@ angular.module('openshiftConsole')
     };
 
     $scope.updateTriggers = function() {
+
+      // Helper for checking webhook objects in the buildConfig spec and creating if missing.
+      function reuseOrCreateWebhook(type) {
+        var webhooks = _.filter($scope.buildConfig.spec.triggers, function(obj) { return obj.type === type})
+        if (_.isEmpty(webhooks)) {
+          webhooks = {
```
Comment (jwforres) [inline assets/app/scripts/controllers/edit/buildConfig.js]: isnt webhooks supposed to be an array?

Comment (jwforres) [inline assets/app/scripts/controllers/edit/buildConfig.js]: or at least a map, not sure which, either way this doesnt seem right

Comment (jhadvig) [inline assets/app/scripts/controllers/edit/buildConfig.js]: @jwforres no just an array of objects.
Check https://paste.fedoraproject.org/358316/14612661/

Comment (jwforres) [inline assets/app/scripts/controllers/edit/buildConfig.js]: if webhooks is supposed to be an array, which looking at the later code seems like it should be, then 1) setting webhooks to {type: type} which is an object isn't right, and 2) webhooks["github"] is also not right since that is not an array reference that is a map reference.

i think all you really want here is

```
if (_.isEmpty(webhooks)) {
  var webhook = {
    type: type,
    secret: ApplicationGenerator._generateSecret()
  }
  webhooks.push(webhook);
```

Comment (jhadvig) [inline assets/app/scripts/controllers/edit/buildConfig.js]: Well `concat()` will take either array or an object and appends it to the array on which its called.
But I see your point and will update
```
var webhook = {
   type: type
};
webhook[(type === "GitHub") ? "github" : "generic"] = {
    secret: ApplicationGenerator._generateSecret()
};
webhooks.push(webhook);
```
since the `secret` needs to be an object of the lowercase webhook type
```
            {
                "type": "GitHub",
                "github": {
                    "secret": "aaed4af14894bc36"
                }
            }
```

Thanks ! :)

**Comment by jhadvig** [comment]
@jwforres updated
